api/hasura/actions/create_circle.ts

```diff
@@ -0,0 +1,221 @@
+import type { VercelRequest, VercelResponse } from '@vercel/node';
+import { GraphQLClient, gql } from 'graphql-request';
+import Validator from 'validatorjs';
+
+export default async function handler(req: VercelRequest, res: VercelResponse) {
+  const variables = req.body.input.object;
+  const rules = {
+    user_name: 'required|string|max:255',
+    address: ['required', 'regex:/^(0x)?[0-9a-f]{40}$/i'],
+    circle_name: 'required|string|max:255',
+    protocol_id: 'integer',
+    protocol_name: 'required_without:protocol_id|string|max:255',
+    uxresearch_json: 'json',
+  };
+
+  Validator.register(
+    'json',
+    function (value) {
+      try {
+        JSON.parse(value);
+      } catch (e) {
+        return false;
+      }
+      return true;
+    },
+    'The :attribute is not a valid JSON.'
+  );
+
+  const validation = new Validator(variables, rules);
+  if (validation.fails()) {
+    return res.status(422).json({
+      extensions: validation.errors.all(),
+      message: 'Invalid input',
+      code: '422',
+    });
+  }
+  const client = new GraphQLClient(process.env.REACT_APP_HASURA_URL, {
+    headers: { 'x-hasura-admin-secret': process.env.HASURA_ADMIN_SECRET },
+  });
+  variables.address = variables.address.toLowerCase();
+  try {
+    // attach new circle to existing organisation
+    if (variables.protocol_id) {
+      const QUERY = gql`
+        query ($address: String!, $protocol_id: Int!) {
+          profiles(
+            limit: 1
+            where: {
+              address: { _eq: $address }
+              users: {
+                role: { _eq: 1 }
+                circle: { protocol_id: { _eq: $protocol_id } }
+              }
+            }
+          ) {
+            id
+          }
+        }
+      `;
+      let data = await client.request(QUERY, {
+        address: variables.address,
+        protocol_id: variables.protocol_id,
+      });
```

Comment:
We should just use GraphQL Zeus here as well so we get strong type safety etc. Can generate an "admin zeus client" separately from the zeus client generated for clientside so it only allows using queries and fields that are allowed for the given role.